python/ray/util/iter.py

```diff
@@ -733,6 +772,9 @@ class LocalIterator(Generic[T]):
 
     thread_local = threading.local()
 
+    UNION_MAX_PULL = 1000
```

Comment:
Before this value hardcoded and not accesible, with a value of 100.
The value controls how many pull operations are going to be tried before stoping (StopIteration) an UNION of PARALLEL-ITERATORS. 
Also, before, all pull reties over an UNION of PARALLEL-ITERATORS were made one after the other without any waiting time. This was a too optimistic approach that could lead to a lot of false-negative when the source of one of the iterators in the UNION is slow (or even worst if all iterators are slow).

Here we are not only increasing the amount of retries to 1000 but also introducing an exponential backoff logic in these retries operations.

There are some trade-off with this approach though:
- FAST sources: the exponential backoff logic could slow down a little the pipeline when some sort of intermittent failure or delays.
- SLOW sources: the default values could be too small if the sources are extremely slow.